File: easy/top-100-liked/n1_two_sum.py
"""
#1. Two Sum (Top 100 Liked Questions)

When constructing a dictionary to store index information,
We have to construct it when we are in the first for loop.

If we attempt to construct a dictionary first,
then try to access the right two indices,
there is a problem of duplicates and things get more complicated.
"""


def twoSum(nums: List[int], target: int) -> List[int]:
    
    loc = {}
    
    for i in range(len(nums)):            
        remainder = target - nums[i]            
        if remainder in loc:
            return [i, loc[remainder]]            
        else:
            loc[nums[i]] = i
            

File: 1-two-sum/n1_two_sum.cpp
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {

        map<int, int> loc;
        int remainder;
        vector<int> output;

        for (int i=0; i < nums.size(); i++) {
            remainder = target - nums[i];

            if (loc.find(remainder) == loc.end()) {
                loc[remainder] = i
            } else {
                output.push_back(loc[remainder]);
                output.push_back(i);
                return output;
            }


        }

    }
};